src/bundler/bundle_v2.zig

```diff
@@ -7546,6 +7548,90 @@ pub const LinkerContext = struct {
+                    }
+
+                    // Require of a top-level await chain is forbidden
+                    if (record.kind == .require) {
+                        var notes = std.ArrayList(Logger.Data).init(c.allocator);
+                        var tla_pretty_path: string = "";
+                        var other_source_index = record.source_index.get();
+
+                        // Build up a chain of notes for all of the imports
+                        while (true) {
+                            const parent_result = c.parse_graph.ast.get(other_source_index);
+                            const parent_source_index = other_source_index;
+
+                            if (parent_result.top_level_await_keyword.len > 0) {
+                                tla_pretty_path = c.parse_graph.input_files.items(.source)[other_source_index].path.pretty;
+                                notes.append(Logger.Data{
+                                    .text = std.fmt.allocPrint(c.allocator, "The top-level await in {s} is here:", .{tla_pretty_path}) catch bun.outOfMemory(),
+                                }) catch bun.outOfMemory();
+                                break;
+                            }
+
+                            if (!Index.isValid(Index.init(parent_result.tla_check.parent))) {
+                                notes.append(Logger.Data{
+                                    .text = "unexpected invalid index",
+                                }) catch bun.outOfMemory();
+                                break;
+                            }
+
+                            other_source_index = parent_result.tla_check.parent;
+
+                            notes.append(Logger.Data{
+                                .text = std.fmt.allocPrint(c.allocator, "The file {s} imports the file {s} here:", .{
+                                    c.parse_graph.input_files.items(.source)[parent_source_index].path.pretty,
+                                    c.parse_graph.input_files.items(.source)[other_source_index].path.pretty,
+                                }) catch bun.outOfMemory(),
+                            }) catch bun.outOfMemory();
+                        }
+
+                        const imported_pretty_path = c.parse_graph.input_files.items(.source)[source_index].path.pretty;
+                        const text: string = if (strings.eql(imported_pretty_path, tla_pretty_path))
+                            std.fmt.allocPrint(c.allocator, "This require call is not allowed because the imported file \"{s}\" contains a top-level await", .{imported_pretty_path}) catch bun.outOfMemory()
+                        else
+                            std.fmt.allocPrint(c.allocator, "This require call is not allowed because the transitive dependency \"{s}\" contains a top-level await", .{tla_pretty_path}) catch bun.outOfMemory();
+
+                        const source: Logger.Source = c.parse_graph.input_files.items(.source)[source_index];
```

Comment:
this does a stack copy and then only references by pointer

let's avoid the stack copy entirely via `&`

Reply:
https://github.com/oven-sh/bun/pull/14218/commits/b6add9c5314fddbdecff3f925a6bd09fd6517e1c